Event listener Docs + Wrapper Types (#333)
webb-tools/gadget: docs/event_listeners.md

## Creating Event Listeners

To create an event listener, begin by defining a struct or enum which will listen to events:

```rust
use gadget_sdk::event_listener::EventListener;
use async_trait::async_trait;

/// Define a simple event listener that ticks every MS milliseconds.
pub struct Ticker<const MS: usize> {
    additional_delay: u64,
}
```

Then, define a context for the event listener:

```rust
#[derive(Copy, Clone)]
pub struct MyContext {
    pub additional_delay: u64,
}
```

Next, implement `EventListener` for `Ticker`:

```rust
/// Implement the [`EventListener`] trait for the Ticker struct. [`EventListener`] has two type parameters:
/// 
/// - T: is the type of the event that the listener listens for. This can be anything that is Send + Sync + 'static.
/// In this case, we are using [`Instant`], which is a timestamp.
///
/// - Ctx: is the context type that the listener receives when constructed. This can be anything that is Send + Sync + 'static,
/// with the special requirement that it is the *first* listed additional parameter in the [`job`] macro (i.e., a parameter not

> /// in params(...)). In this case, we are using [`MyContext`].

Reviewer: That seems like an odd restriction? Why not mark it with `#[context]` or check for a parameter named `context`?

Author: I would also like to find an improvement over this. For now, this is just what was agreed upon by @shekohex and I.

Reviewer 2: We could go with the more explicit path here with adding `#[context]` but we are aiming for less noise which would improve readability.

```rust
#[job(id = 1, params(x, y), result(_)]
fn my_job(context: Context, x: u64, y: u64) -> Result<u64, Inflatable> { ... } 
```
vs
```rust
#[job(id = 1, params(x, y), result(_)]
fn my_job(#[context] context: Context, x: u64, y: u64) -> Result<u64, Inflatable> { ... } 
```

    {
        Ok(Self { additional_delay: context.additional_delay })
    }

    /// Implement the logic that looks for events. In this case, we have a simple stream
    /// that returns after MS milliseconds.
    async fn next_event(&mut self) -> Option<Instant> {
        tokio::time::sleep(tokio::time::Duration::from_millis(MS as u64 + self.additional_delay)).await;
        Some(Instant::now())
    }

    /// After next_event is called, the event gets passed here. This is where you would implement
    /// listener-specific logic.
    async fn handle_event(&mut self, _event: Instant) -> Result<(), Error> {
        Ok(())
    }
}
```

Finally, register the event listener inside the `job` macro using `event_listener`:

```rust
#[job(
    id = 0,
    params(x),
    result(_),
    event_listener(Ticker::<6000>), // <-- Register the event listener here
    verifier(evm = "IncredibleSquaringBlueprint")
)]
pub fn hello_event_listener(
    x: u64,
    context: MyContext, // <-- The context type must be the first additional parameter
    env: GadgetConfiguration<parking_lot::RawRwLock>,
) -> Result<u64, Infallible> {
    Ok(x.saturating_pow(2u32))
}
```

## Wrapper Event Listeners

> Various wrappers exist to simplify how common operations are performed. For example, there is the `PeriodicEventListener`:

Reviewer: ```suggestion
Various wrappers exist to simplify how common operations are performed.
```

### PeriodicEventListener
Some programs may only be interested in checking for events at regular intervals. In this case, the `PeriodicEventListener` can be used to simplify the process.

A `PeriodicEventListener` is a wrapper that takes 4 type parameters:

> * `MSEC`: The number of milliseconds between each event check.

Reviewer: nit: The `Ticker` example above used `MS`, this one uses `MSEC`

Author: Adjusted the example above to use `MSEC` so it's all consistent.

* `T`: The inner event listener type

> * `Evt`: The event type

Reviewer: @tbraun96 FYI this is just updated in the doc, should be changed on the struct too.

* `Ctx`: The context type

> We can make a `PeriodicEventListener` that ticks every 5000ms to check the status of a web server
> 
> ```rust
> use reqwest;

Reviewer: ```suggestion
We can make a `PeriodicEventListener` that ticks every 5000ms to check the status of a web server using [reqwest](crates.io/crates/reqwest).

```rust
```
